code/game/machinery/scan_gate.dm

```diff
@@ -0,0 +1,195 @@
+#define SCANGATE_GUNS 			"Guns"
+#define SCANGATE_WANTED			"Wanted"
+#define SCANGATE_SPECIES		"Species"
+
+
+/obj/machinery/scanner_gate
+	name = "scanner gate"
+	desc = "A gate able to perform mid-depth scans on any organisms who pass under it."
+	icon = 'icons/obj/TODO.dmi'
+	icon_state = "scanner_gate"
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 50
+	circuit = /obj/item/circuitboard/machine/scanner_gate
+	
+	var/locked = FALSE
+	var/scangate_mode = SCANGATE_NONE
+	var/disease_threshold = DISEASE_SEVERITY_MINOR
+	var/detect_species = /datum/species/human
+	var/reverse = FALSE //If true, signals if the scan returns false
+	
+/obj/machinery/scanner_gate/examine(mob/user)
+	..()
+	if(locked)
+		to_chat(user, "<span class='notice'>The control panel is ID-locked. Swipe a valid ID to unlock it.</span>")
+	else
+		to_chat(user, "<span class='notice'>The control panel is unlocked. Swipe an ID to lock it.</span>")
+	
+/obj/machinery/scanner_gate/Crossed(atom/movable/AM)
+	..()
+	if(!(stat & (BROKEN|NOPOWER)) && isliving(AM))
+		perform_scan(AM)
+		
+/obj/machinery/scanner_gate/attackby(obj/item/W, mob/user, params)
+	var/obj/item/card/id/card = W.GetID()
+	if(card)
+		if(locked)
+			if(allowed(user))
+				locked = FALSE
+				req_access = list()
+				to_chat(user, "<span class='notice'>You unlock [src].</span>")
+		else if(!(obj_flags & EMAGGED))
+			to_chat(user, "<span class='notice'>You lock [src] with [W].</span>")
+			var/list/access = W.GetAccess()
+			req_access = access
+			locked = TRUE
+		else
+			to_chat(user, "<span class='warning'>You try to lock [src] with [W], but nothing happens.</span>")
+	else
+		return ..()
+			
+/obj/machinery/scanner_gate/emag_act(mob/user)
+	if(obj_flags & EMAGGED)
+		return
+	locked = FALSE
+	req_access = list()
+	obj_flags |= EMAGGED
+	to_chat(user, "<span class='notice'>You fry the ID checking system.</span>")
+	
+/obj/machinery/scanner_gate/proc/perform_scan(mob/living/M)
+	var/beep = FALSE
+	switch(scan_type)
+		if(SCANGATE_NONE)
+			return
+		if(SCANGATE_WANTED)
+			var/perpname = M.get_face_name(M.get_id_name())
+			var/datum/data/record/R = find_record("name", perpname, GLOB.data_core.security)
+			if(!R || (R.fields["criminal"] == "*Arrest*"))
+				beep = TRUE
+		if(SCANGATE_MINDSHIELD)
+			if(M.has_trait(TRAIT_MINDSHIELD))
+				beep = TRUE
+		if(SCANGATE_NANITES)
+			if(SEND_SIGNAL(M, COMSIG_HAS_NANITES))
+				beep = TRUE
+		if(SCANGATE_DISEASE)
+			if(iscarbon(M))
+				var/mob/living/carbon/C = M
+				if(get_disease_severity_value(C.check_virus()) >= get_disease_severity_value(disease_threshold))
+					beep = TRUE
+		if(SCANGATE_SPECIES)
+			if(ishuman(M))
+				var/mob/living/carbon/human/H = M
+				if(is_species(H, detect_species))
+					beep = TRUE
+				if(detect_species == /datum/species/zombie) //Can detect dormant zombies
+					if(H.getorganslot(ORGAN_SLOT_ZOMBIE))
+						beep = TRUE
+		if(SCANGATE_GUNS))
+			for(var/I in M.get_contents())
+				if(istype(I, /obj/item/gun))
```

Comment:
This will go off with wands and staves, blacklist /obj/item/gun/magic?

Reply:
I thought it would be a lot of extra processing to do so, so i kept it simple

Reply:
You could create a static typecache of /obj/item/gun, then subtract `typesof(/obj/item/gun/magic)` from the typecache.

Reply:
that could work, it might allow me to make other types of contraband scans